File: guides/openai-assistant-py/openai_assistant/assistant.py
from typing import List

from dotenv import load_dotenv
import openai
from openai.types.beta.assistant_create_params import Tool

load_dotenv()


def create_assistant():
    client = openai.Client()

    functions: List[Tool] = [
        {
            "type": "function",
            "function": {
                "name": "create_directory",
                "description": "Create a directory",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "directory": {
                            "type": "string",
                            "description": "The path to the directory to be created",
                        },
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "save_content_to_file",
                "description": "Save content (code or text) to file",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "content": {
                            "type": "string",
                            "description": "The content to save",
                        },
                        "path": {
                            "type": "string",
                            "description": "The path to the file, including extension",
                        },
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "list_files",
                "description": "List files in a directory",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "path": {
                            "type": "string",
                            "description": "The path to the directory",
                        },
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "read_file",
                "description": "Read a file",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "path": {
                            "type": "string",
                            "description": "The path to the file",
                        },
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "commit",
                "description": "Commit changes to the repository",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "commit_message": {
                            "type": "string",
                            "description": "The commit message",
                        },
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "make_pull_request",
                "description": "Create a pull request",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "title": {
                            "type": "string",
                            "description": "The title of the pull request",
                        },
                        "body": {
                            "type": "string",
                            "description": "The description or body of the pull request",
                        },
                    },
                },
            },
        },
    ]

    ai_developer = client.beta.assistants.create(
        instructions="""You are an AI developer. You help user work on their tasks related to coding in their codebase. The provided codebase is in the /home/user/repo.
    When given a coding task, work on it until completion, commit it, and make pull request.

    If you encounter a problem, communicate it promptly, please. 

    You can create and save content (text or code) to a specified file (or create a new file), list files in a given directory, read files, commit changes, and make pull requests. Always make sure to write the content in the codebase.

    By default, always either commit your changes or make a pull request after performing any action on the repository. This helps in reviewing and merging your changes.
    Always make the pull request into new branch, and name the pull request "Pull request from AI Developer." Describe the changes in the pull request body based on the content.

    Be professional, avoid arguments, and focus on completing the task.

    When you finish the task, always provide the link to the original repository (not to a particular commit, but to the repo as a whole.)
    Additionally, be prepared for discussions; not everything user writes implies changes to the repository. For example, if the user writes "thank you", you can simply answer "you are welcome".
    But by default, if you are assigned a task, you should immediately do it in the provided repository, and not talk only talk about your plan.
    """,
        name="AI Developer",
        tools=functions,
        model="gpt-4-1106-preview",
    )

    print("AI Developer Assistant created, copy its id to .env file:")
    print(ai_developer.id)


if __name__ == "__main__":
    create_assistant()

File: guides/ai-github-developer-ts/ai_github_developer/assistant.ts
import dotenv from 'dotenv';
import OpenAI from 'openai';
import { REPO_DIRECTORY } from './main.ts';

dotenv.config();



async function createAssistant() {
    const client = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY
    });


    try {
        const aiDeveloper = await client.beta.assistants.create({
            instructions: `You are an AI developer. You help user work on their tasks related to coding in their codebase. The provided codebase is in the ${REPO_DIRECTORY}.
            When given a coding task, work on it until completion, commit it, and make pull request.
        
            If you encounter a problem, communicate it promptly, please.
        
            You can create and save content (text or code) to a specified file (or create a new file), list files in a given directory, read files, commit changes, and make pull requests. Always make sure to write the content in the codebase.
        
            By default, always either commit your changes or make a pull request after performing any action on the repo. This helps in reviewing and merging your changes.
            Name the PR based on the changes you made.
        
            Be professional, avoid arguments, and focus on completing the task.
        
            When you finish the task, always provide the link to the pull request you made (if you made one.)
            Additionally, be prepared for discussions; not everything user writes implies changes to the repo. For example, if the user writes "thank you", you can simply answer "you are welcome".
            But by default, if you are assigned a task, you should immediately do it in the provided repo, and not talk only talk about your plan.`,
            name: "AI Developer",
            tools: [
                {
                    type: "function",
                    function: {
                        name: "createDirectory",
                        description: "Create a directory",
                        parameters: {
                            type: "object",
                            properties: {
                                path: {
                                    type: "string",
                                    description: "The path to the directory to be created",
                                },
                            },
                            required: ["path"],
                        },
                    },
                },
                {
                    type: "function",
                    function: {
                        name: "saveContentToFile",
                        description: "Save content (code or text) to file",
                        parameters: {
                            type: "object",
                            properties: {
                                content: {
                                    type: "string",
                                    description: "The content to save",
                                },
                                path: {
                                    type: "string",
                                    description: "The path to the file, including extension",
                                },
                            },
                            required: ["content", "path"],
                        },
                    },
                },
                {
                    type: "function",
                    function: {
                        name: "listFiles",
                        description: "List files in a directory",
                        parameters: {
                            type: "object",
                            properties: {
                                path: {
                                    type: "string",
                                    description: "The path to the directory",
                                },
                            },
                            required: ["path"],
                        },
                    },
                },
                {
                    type: "function",
                    function: {
                        name: "readFile",
                        description: "Read a file",
                        parameters: {
                            type: "object",
                            properties: {
                                path: {
                                    type: "string",
                                    description: "The path to the file",
                                },
                            },
                            required: ["path"],
                        },
                    },
                },
                {
                    type: "function",
                    function: {
                        name: "commit",
                        description: "Commit changes to the repo",
                        parameters: {
                            type: "object",
                            properties: {
                                commitMessage: {
                                    type: "string",
                                    description: "The commit message",
                                },
                            },
                            required: ["commitMessage"],
                        },
                    },
                },
                {
                    type: "function",
                    function: {
                        name: "makePullRequest",
                        description: "Creates a new branch and makes a pull request",
                        parameters: {
                            type: "object",
                            properties: {
                                title: {
                                    type: "string",
                                    description: "The title of the pull request",
                                }
                            },
                            required: ["title"],
                        },
                    },
                },
            ],
            model: "gpt-4-1106-preview",
        });

        console.log("AI Developer Assistant created, copy its id to .env file:");
        console.log(aiDeveloper.id);

    } catch (error) {
        console.error('Error creating AI Developer Assistant:', error);
    }
}

createAssistant().catch(error => console.error(error));
